reactor-netty-http/src/test/java/reactor/netty/http/client/HttpClientTest.java

```diff
@@ -3401,4 +3403,46 @@ static void testIssue3285SendRequest(HttpClient client, @Nullable Class<? extend
 			        .verify(Duration.ofSeconds(5));
 		}
 	}
+
+	@Test
+	void testIssue3416() {
+		disposableServer =
+				createServer()
+				        .route(r -> r.get("/", (req, res) -> res.sendString(Mono.just("testIssue3416")))
+				                     .ws("/ws", (in, out) -> out.neverComplete()))
+				        .bindNow();
+
+		AtomicReference<WeakReference<Connection>> connWeakRef = new AtomicReference<>();
+		HttpClient client =
+				createClient(disposableServer.port())
+				        .observe((conn, state) -> {
+				            if (state == ConnectionObserver.State.CONNECTED) {
+				                connWeakRef.compareAndSet(null, new WeakReference<>(conn));
+				            }
+				        });
+
+		client.get()
+		      .uri("/")
+		      .response() // Reactor Netty will close the connection
+		      .flatMap(res ->
+		          client.websocket()
+		                .uri("/ws")
+		                .handle((in, out) ->
+		                    Flux.range(0, 10)
+		                        .delayElements(Duration.ofMillis(100))
+		                        .skipUntil(l -> {
+		                            boolean result = connWeakRef.get().get() == null;
+		                            if (!result) {
+		                                System.gc();
+		                            }
+		                            return result;
+		                        })
+		                        .switchIfEmpty(Mono.error(new RuntimeException("failed"))
+		                        .flatMap(l -> Mono.empty())))
+		                .then()
+		                .contextWrite(Context.of(res.getClass(), res)))
```

Comment:
I would love to see a comment that this serves the purpose of addressing another concern. While the main test flow validates that the connection is freed once response is retrieved as `res` reference, the `contextWrite` store ensures also that the res in `Context` also doesn't preserve the `Connection` reference. Am I understanding it right?

Reply:
correct